pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -171,6 +211,140 @@ func referencesDiffs(old []metatypes.OwnerReference, new []metatypes.OwnerRefere
 	return added, removed
 }
 
+func shouldOrphanDependents(e event, accessor meta.Object) bool {
+	// The delta_fifo may combine the creation and update of the object into one
+	// event, so we need to check AddEvent as well.
+	if e.oldObj == nil {
+		if accessor.GetDeletionTimestamp() == nil {
+			return false
+		}
+	} else {
+		oldAccessor, err := meta.Accessor(e.oldObj)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("cannot access oldObj: %v", err))
+			return false
+		}
+		// ignore the event if it's not updating DeletionTimestamp from non-nil to nil.
+		if accessor.GetDeletionTimestamp() == nil || oldAccessor.GetDeletionTimestamp() != nil {
+			return false
+		}
+	}
+	return sets.NewString(accessor.GetFinalizers()...).Has(api.OrphanFinalizer)
+}
+
+func (gc *GarbageCollector) orhpanDependents(owner *node, dependents []*node, retries int) error {
```

Comment:
> do you need to pass in dependents? Can't you get that from the owner?

I'm passing in copied of the points of the owner's dependents, so that I don't need to lock them.

> should owner be locked?

I changed to pass in owner.identity, which doesn't need to be locked.